### ISSUE TITLE
How to share data between different operators

### ISSUE THREAD

**Issue by gyliu513** [operator-framework/operator-lifecycle-manager#1293]
I have two operators, one is IAM and another is MongoDB, and IAM need to depend on some parameters from MongoDB operator, like MongoDB service URL etc. Does there are any best practise for how I can share those data of MongoDB to IAM operator?

I have two solutions now:

## Solution 1

![image](https://user-images.githubusercontent.com/4461983/74639364-bc13ef00-51a8-11ea-9b25-d0eb28068ba7.png)

Enable MongoDB opertor generate some configMap or secrets and let IAM operator get the shared data. The problem for this solution is that I need to let IAM operator know the namespace of the configMap and secrets and also need to have permission to access it.

## Solution 2

![image](https://user-images.githubusercontent.com/4461983/74639542-0ac18900-51a9-11ea-8a8a-3fa41d219943.png)

This option is enabling IAM create a CR resource named as `SharedKlusterInfoRequest`, and MongoDB operator will help to reconcile this CR and inject the shared info to the namespace of IAM operator and then IAM operator will be able to get those data.

Any comments or suggestions for this? Thanks.

**Comment by pgodowski** [comment]
@gyliu513 isn't it that with _Solution 2_, Mongo would need to have cluster-level view and then write access to the individual consumers' (like IAM's) namespaces?

Why can't we have all the components in the single namespace, i.e. what is the advantage of having those components each living in different namespaces? Customers asked for moving off from `kube-system`, but not (yet?) asked for creating multiple namespaces for _Common Services_, which potentially might increase the operational management cost, a little complicate consolidated usage tracking.

**Comment by hassenius** [comment]
This problem does not seem completely dissimilar to the problem faced by the [cloud-credential-operator](https://github.com/openshift/cloud-credential-operator).

The implementation is pretty much like your Solution 2.

In this implementation model the requesting service (your IAM) creates a CR specifying what data is needed and where the data should be stored.

The operator knowing the desired information (your MongoDB operator) will go ahead and store the requested data (as configmap or secret) at the relevant location (namespace).

To @pgodowski point:

Of course the location (configmap/secret in namespace) needs to be accessible by both the requester (IAM) and the fulfiller (MongoDB). It seems naturally that this location is the IAMs namespace, and that the relevant RBAC to allow MongoDB access the location is configured by IAM Operator during the existing setup/preparation procedure.
In other words, the MongoDB does not need to have cluster level access, but rather can be granted fine grained permissions by services that require it.
 
As pointed out by this discussion, one important benefit of having multiple namespaces over a single namespace is to limit the blast radius of a compromise, since namespaces creates a logical boundary for configmaps and secrets.

**Comment by jianzhangbjz** [comment]
@ecordell What do you suggest? Thanks!

**Comment by dinhxuanvu** [comment]
This issue is stale. Closed. Please feel free to reopen if needed.
